_posts/2019-01-14-code-of-the-week.md

```diff
@@ -0,0 +1,127 @@
+to validate the input of variables. Since the introduction of datatypes, we can
+enforce the correct datatype directly when the user provides data, without
+calling multiple functions. The datatypes are especially useful for
+[Collection datatypes](https://puppet.com/docs/puppet/6.1/lang_data_abstract.html#the-collection-data-type)
+because it's now possible to very strict types for each element in a
+collection. The migration to datatypes was a bit painful
+[in](https://github.com/puppetlabs/puppetlabs-apache/pull/1621)
+[some](https://github.com/puppetlabs/puppetlabs-postgresql/pull/852)
+[modules](https://github.com/voxpupuli/puppet-nginx/pull/1056). But the result
+was nice. Not only were the actual code
+([LOC](https://en.wikipedia.org/wiki/Source_lines_of_code)) and tests shorter,
+datatypes also allow us to validate way more in detail than the old validate
+functions. This reduced the amount of possible errors that could happen because
+of unexpected input. We actually discovered multiple bugs during the transition
+to datatypes and also many codepieces that were suboptimal. For example a lot
+of casting from `'true'` (a string, valid input data in some modules) to `true`
+(as an actual
+[Boolean](https://puppet.com/docs/puppet/6.1/lang_data_boolean.html) datatype).
+Based on the experiences we made, we created [review guidelines and documented
+best practices](https://voxpupuli.org/docs/#reviewing-a-module-pr).
+
+Some of these guidelines are:
+
+> Are new parameters introduced? They must have datatypes
+
+> Are datatypes from stdlib used? Ensure that lowest supported stdlib version is 4.13.1. Check if a newer version introduced the used datatype
+
+> Are there new params with datatype Hash or Array? If possible, they should default to empty Hash/Array instead of undef. You can also enforce the datastructure like Array[String]
+
+> Are there new params with datatype Boolean? The default value is a tricky decision which needs careful reviewing. Sometimes a True/False is the better approach, sometimes undef
+
+> Does a new param map to an option in a config file of a service? The Parameter should accept the possible values that the service allows. For example `'on'` and `'off'`. Don’t accept a boolean that will be converted to `'on'` or `'off'`
+
+> If you can supply one or multiple values for an attribute it’s common practice to enforce the datatype for one value and an array of that datatype. An example for string is `Variant[String[1],Array[String[1]]]`. This can be used in the Puppet code as `[$var].flatten()`
+
+**The summary: Datatypes are useful and important. Always using them for each
+exposed parameter will help you catch bugs earlier, and the more detailed
+your datatype is the better it will be at catching bugs.**
+
+After a very long introduction, let us come to the actual contribution. In
+calender week three we have a contribution from [David
```

Comment:
calendar

Reply:
lol, `calender` is an actual word, but has a totally different meaning.